travis: adding debian and opensuse CI tests

c-ney opened the pull request:
**Description of the Pull Request (PR):**

This is a follow up (or more like a replacement) to #1204. It includes some refactoring of the CI stuff and introduces debian 8/9 (by @olifre with some commit amendments by me, hope this is okay) and opensuse 42.3/15.0 tests. The latter are not enabled for travis yet and still somewhat of a WIP. Running locally, both opensuse versions currently fails at `39-deffile.sh` due to the issue from #1074.

The first commit for autoconf is identical to #1712 (didn't rebase yet). The debian changelog "fix" should likely go into a separate issue/pr.

In `centos_run` --now renamed to `docker_run`-- I did not fully understand the implications of `/usr/sbin/init` for centos 7. Debian and opensuse do not have it, so I just did some copy&paste from centos 6 and removed the `centos6_setup` line. Would be great if someone could elaborate.

Thanks for any feedback!

**This fixes or addresses the following GitHub issues:**

- Ref: #


**Checkoff for all PRs:**

- [x] I have read the [Guidelines for Contributing](https://github.com/singularityware/singularity/blob/master/CONTRIBUTING.md), and this PR conforms to the stated requirements.
- [x] I have added changes to the [CHANGELOG](https://github.com/singularityware/singularity/blob/development/CHANGELOG.md) and and documentation updates to the [singularityware](https://www.github.com/singularityware/singularityware.github.io) documentation base.
- [x] I have tested this PR locally with a `make test`
- [x] This PR is NOT against the project's master branch
- [x] I have added myself as a contributor to the [contributors's file](https://github.com/singularityware/singularity/blob/master/CONTRIBUTORS.md)
- [x] This PR is ready for review and/or merge


Attn: @singularityware-admin

olifre committed:
travis: Add testing and packaging test for Debian 8 and 9.

olifre committed:
travis: Add test that distclean target works.

olifre committed:
travis: Specify --create-home when creating testuser.

On many distributions, creation of the home directory
is not handled by useradd by default.

c-ney committed:
debian/changelog: fix missing revision number

debian/source/format is "3.0 (quilt)" or "non-native", thus a revision
number is required

c-ney committed:
travis: rename centos_run to be more generic

c-ney committed:
travis: make starting build tests more generic

c-ney committed:
travis: move distribution specific dependencies...

and commands into separate files.

c-ney committed:
travis: Fix final spec file not existing...

before configure.

We want to use the spec file to install build dependencies for the
configure step, but we have to run configure to actually finalize the
file. At this point only the package version is changed by autotools, we
can do this manually as well.

c-ney committed:
travis: Install newest pylint version from pypi

Some distributions (notably centos 6, debian 8 and opensuse 42.3) ship with
old/buggy versions of pylint3 in the repositories. Use pip3 to get the
newest version available.

c-ney committed:
travis: enable debian tests

c-ney committed:
travis: add opensuse tests

c-ney committed:
travis: leap 15.0 container lacks hostname command

jmstover commented on .travis/debuild_test:
```diff
@@ -0,0 +1,48 @@
+
+echo "Directory containing the build .deb-files:"
+ls -la ..
+gdebi -n ../singularity-container_*.deb
```
Instead of adding `gdebi` ... could you just use `dpkg -i` to install the `.deb` file?

olifre replied:
`dpkg` does not resolve dependencies, while `gdebi` does. So it's essential to use `gdebi` here to check that the dependencies expressed in the packaging scripts are actually complete.

olifre replied:
> dpkg does not resolve dependencies, while gdebi does.

Or rather: `dpkg` _does_ resolve dependencies, of course, but does not install them (it only complains if things are not installed yet) :wink:.

c-ney replied:
I just found that you can apparently do `dpkg -i pkg.deb` followed by `apt-get -fy install` [1]. Will try this later.

[1] https://superuser.com/a/537371

jmstover replied:
If you're going to do that, couldn't you do:

```
apt-get install ../singularity-container_*.deb
```

And then let `apt-get` pull the dependencies?

olifre replied:
Indeed, both of these should also work - of course `apt-get -fy install` may do more than wanted / expected, and `apt-get install ../singularity-container_*.deb` is an officially undocumented feature. The man-page says `Each package is a package name, not a fully qualified filename` when talking about parameters to `install`. So this unofficial feature may be broken in any future version of `apt-get`.

jmstover replied:
Hrmm.... If that's the case, then I guess I would stick with installing `gdebi` and using it, since the `apt-get -fy install` may have unintended consequences, and `apt-get install package_name.deb` is a unsupported "feature."

I would have just thought apt-get could install from a local `.deb` package file as if it was pulling from sources.list, and not thought anything of it.

jmstover commented on .travis/debuild_test:
```diff
@@ -0,0 +1,48 @@
+apt-get -y install libtool sudo e2fsprogs
+sed -i 's,^prefix=.*,prefix=/usr,' test.sh
+sed -i 's,^sysconfdir=.*,sysconfdir=/etc,' test.sh
+sed -i 's,^localstatedir=.*,localstatedir=/var,' test.sh
```
```
prefix=
exec_prefix=
libexecdir=
sysconfdir=
localstatedir=
bindir=
```

Should all be set to what was used on the `./configure` line (`test.sh` gets generated from `test.sh.in`). There's hard coded paths that need to match up here because of the SUID bits.

olifre replied:
Oh yes! It seems this is not needed anymore. As of #1204, when I added those lines last year, `test.sh` was not correctly taking over what was passed to `configure`, but it seems that's corrected now. @c-ney : Just removing those `sed` lines should work :wink:.

c-ney replied:
Now that you mention this, I remember having to put in an opensuse specific line in the past, but forgot to add it here and it still works. Will remove them from ~~both~~ `debuild_test` ~~and `rpmbuild_test`~~

Edit: No, this doesn't work for the rpm test due to different `./configure` arguments in the spec file...

olifre replied:
Ahh! I remember why this was needed! It works fine on Debian, since `debuild`runs "in-place" and executes `configure` with full parameters. 
`rpmbuild`, though, runs in a sandbox, and calls `configure` there, and not where we actually started it. 
So either we have to steal `test.sh` from `/root/rpmbuild/BUILD/singularity-2.5.1.999/`, or replicate the parameters `rpmbuild` is using further up (i.e. before `make dist`) when `configure` is called to make `make dist` actually work. 
I'm unsure which is nicer...

jmstover replied:
Right... and this is on the debian one... I'm working on a comment on the rpmbuild one. ;)

jmstover commented on .travis/rpmbuild_test:
```diff
@@ -2,31 +2,41 @@
 sed -i 's,^sysconfdir=.*,sysconfdir=/etc,' test.sh
 sed -i 's,^localstatedir=.*,localstatedir=/var,' test.sh
-useradd -u 1000 testuser
+useradd -u 1000 --create-home testuser
```
On these sed lines, you should use the values from:
```
rpm -E '%{_prefix}'
rpm -E '%{_sysconfdir}'
rpm -E '%{_localstatedir}'
```

That will pull the expanded macro value out in case it ever changes... for some reason.

*Edit:*
So something like:
```
sed -i "s@^prefix=.*@prefix=$(rpm -E '%{_prefix}')@" test.sh
```

olifre replied:
Good idea! I'm all for it :+1:

c-ney replied:
I also stumbled upon this:

```sh
bash -c "$(rpm --eval '%configure')"
make dist
```

...but I don't think anyone would configure like this before a make dist... so, whatever you prefer!

jmstover replied:
I've ... seen it. Usually I've needed to rip out the normal `%configure` macro, and use my own options, and/or override the normal `_libexecdir`, `_prefix`, etc... 

But, yeah... I've sadly seen that before. :/

c-ney committed:
travis: don't modify test.sh in debuild_test

c-ney committed:
travis: adjust test.sh for %configure rpm macro

inside the rpmbuild sandbox, the %configure macro gets expanded to set
distribution specific directories in the configure step. These diverge
from the ones set before `make dist` on the "outside", so we have to
adjust test.sh accordingly.

jmstover reviewed (approved):
This all appears good...

GodloveD reviewed (changes requested):
This looks good to me.  Just a few asks.  Please add your name to the contributor's file @c-ney, and document these upgrades in the changelog under 2.6.0.  Thanks!

c-ney committed:
travis: add news to changelog

c-ney committed:
travis: add myself to contributors list

c-ney committed:
travis: enable opensuse build tests

c-ney committed:
travis: fix missing /etc/localtime for opensuse

the opensuse leap docker container are missing /etc/localtime by
default, thus producing the warning

WARNING: Non existent 'bind path' source: '/etc/localtime'

which then breaks the

stest 0 cmp "$SINGULARITY_TESTDIR/actual.txt" "$SINGULARITY_TESTDIR/expected.txt"

in tests/39-deffile.sh.

c-ney committed:
debian/changelog: dirty fix adding maintainer

dpkg-buildpackage: error: unable to determine source changed by

GodloveD reviewed (approved):
lgtm.  @jmstover do you like these new changes too?

jmstover commented on debian/changelog:
```diff
@@ -1,12 +1,14 @@
-singularity-container (2.5.1) unstable; urgency=high
+singularity-container (2.5.1-1) unstable; urgency=high
```
Does the revision (i.e. the `-1`) *need* to exist? I have done a build on Debian 8/9 and without putting one in it defaulted to a `-1` revision without. I think...  Pretty sure I didn't make any changes my first few build attempts.

IIRC, there was also a comment made ages ago that the revision didn't need to be put there as the package maintainer would be modifying it...

c-ney replied:
At least in the default configuration, building a debian package breaks at the `dpkg-source` step inside `debuild`. Using this container

```
bootstrap: docker
from: debian:8

%environment
    LC_ALL=C

%post
    apt-get update
    apt-get -y dist-upgrade
    apt-get -y install devscripts build-essential lintian gdebi-core git
```

as the environment gives me

```
Singularity deb8.simg:~/devel/extern/singularity> git status
On branch development-2.x
Your branch is up-to-date with 'origin/development-2.x'.
nothing to commit, working directory clean
Singularity deb8.simg:~/devel/extern/singularity> dpkg-source -b .
parsechangelog/debian: warning: singularity_clean/debian/changelog(l9): found start of entry where expected more change data or trailer
LINE: singularity-container (2.5.0) unstable; urgency=high
parsechangelog/debian: warning: singularity_clean/debian/changelog(l9): found eof where expected more change data or trailer
dpkg-source: error: can't build with source format '3.0 (quilt)': non-native package version does not contain a revision
```

This goes away when you add the revision and the maintainer.

```
Singularity deb8.simg:~/devel/extern/singularity> dpkg-source -b .
dpkg-source: info: using source format `3.0 (quilt)'
dpkg-source: info: building singularity-container using existing ./singularity-container_2.5.1.orig.tar.gz
dpkg-source: info: building singularity-container in singularity-container_2.5.1-1.debian.tar.xz
dpkg-source: info: building singularity-container in singularity-container_2.5.1-1.dsc
```

It all comes down to what source format is set in `debian/source/format`. `3.0 (quilt)` for non-native wants the revision, `3.0 (native)` does not.

I'm no expert in debian package building, so maybe this is a non-issue due to some convention as you mentioned.

Related PR: #1736

GodloveD replied:
I'm fine with this.  Either way the maintainer has to edit the file.

GodloveD committed:
Merge branch 'development-2.x' into travis-again
